simulator/src/models/power-plant.ts

```diff
@@ -0,0 +1,564 @@
+
+/***************************************************************************
+ * The power-plant model holds the neccessary parameters and power-plant data.
+ ***************************************************************************/
+
+
+import Simulation from "../simulation";
+import Battery from "./battery";
+import uuid from "uuid";
+import { ElectricityGridDB, eq } from "./database";
+import { randomFloat }  from "./utils";
+
+
+/**
+ * PowerPlantInfo is a data structure for holding power plant information.
+ */
+export interface PowerPlantInfo {
+    readonly id: uuid.v4;
+    readonly time: Date;
+    readonly status: Status;
+
+    readonly productionLevel: number;
+    readonly productionCapacity: number;
+    readonly productionVariant: number;
+    readonly productionRatio: number;
+
+    readonly batteryValue: number;
+    readonly batteryCapacity:number;
+
+    readonly totalProduction: number;
+
+    readonly unit: string;
+}
+
+
+/**
+ * The power-plant model holds the parametes and model used to generate electricity.
+ * Different power-plant models can be used to simulate different power-plants.
+ */
+export default class PowerPlant {
+    /**
+     * The uuid of the power plant.
+     */
+    private _id: uuid.v4;
+
+    /**
+     * The status of the power plant.
+     */
+    private _status: Status;
+
+    /**
+     * The time it takes for the pwoer plant to start.
+     */
+    private _startDelay: number;
+
+    /**
+     * The time it takes for the pwoer plant to stop.
+     */
+    private _stopDelay: number;
+
+    /**
+     * The number of kw that should be produced by the power-plant. 
+     */
+    private _productionLevel: number;
+    
+    /**
+     * The number of kw that the power plant max can produce.
+     */
+    private _productionCapacity: number;
+
+    /**
+     * The number of kw that can differ from the productionLevel.
+     */
+    private _productionVariant: number;
+
+    /**
+     * The percent of produced electricity that is sent to the market.
+     */
+    private _productionRatio: number;
+
+    /**
+     * The power plants battery.
+     */
+    private _battery: Battery;
+
+    /**
+     * The uuid of the power plant owner.
+     */
+    private _owner: uuid.v4;
+
+    /**
+     * The unit.
+     */
+    private _unit: string;
+
+    /**
+     * Keep track of the time when the last wind speed was generated.
+     */
+    private _time: Date;
+    
+    /**
+     * The timestamp when this wind object was created.
+     */
+    public _createdAt: Date;
+
+    /**
+     * The timestamp when the wind simulation was last updated.
+     */
+    private _updatedAt: Date;
```

Comment:
Does all these fields need to have getters there are maybe some variables that are only used inside this particular class e.g. producitonCapacity, productionVariant, battery, unit, time does not seem that useful to anyone else outside this class except battery but that should maybe not be accessed directly but via a method. Also _createdAt should be private.

Reply:
Note; if there is a value constraint for a field e.g. productionLevel cannot be less than 0 then it makes sense to use setter to prevent that. Reconsider if everything really needs to be accessible outside this class.

Reply:
Most of the variables need to be accessible because we need to display them on to the manager. The note I don't understand because there is setters for variables like productionLevel, what do you mean there?

Reply:
Yes but getPowerPlantInfo() already gives you all the information you need. I think this is enough to create the REST API for displaying information.